lisp/ess-r-mode.el

```diff
@@ -288,6 +288,50 @@ namespace.")
      (t
       font-lock-comment-face))))
 
+(defvar ess-r--non-fn-kwds
+  '("in" "else" "break" "next" "repeat"))
+
+(defvar-local ess-r--keyword-regexp nil)
+(defun ess-r--find-fl-keyword (limit)
+  "Search for R keyword and set the match data.
+To be used as part of `font-lock-defaults' keywords."
+  (unless ess-r--keyword-regexp
+    (let (fn-kwds non-fn-kwds)
+      (dolist (kw ess-R-keywords)
+        (if (member kw ess-r--non-fn-kwds)
+            (push kw non-fn-kwds)
+          (push kw fn-kwds)))
+      (setq ess-r--keyword-regexp
+            (concat "\\("
+                    (regexp-opt non-fn-kwds 'words)
+                    "\\)\\|\\("
+                    (regexp-opt fn-kwds 'words)
+                    "\\)"))))
+  (let (out)
+    (while (and (not out)
+                (re-search-forward ess-r--keyword-regexp limit t))
+      (let ((data (match-data)))
+        (setq out (if (match-beginning 1)
+                      (pcase (match-string 1)
+                        ;; Only fontify `in` if part of a `for` construct
+                        (`"in"
+                         (save-excursion
+                           (goto-char (match-beginning 1))
+                           (and (ess-backward-up-list)
+                                (forward-word -1)
+                                (looking-at "for\\s-*("))))
+                        ;; Remaining non-function-like keywords: Fontify
```

Comment:
You're right I don't know that we will. Even if we get the robust expression matcher (I'll get back to work at it for the next release cycle), it might not be efficient enough for font-lock. Regarding efficiency I think I'll change the tokenizer so that it returns pure symbols instead of cons cells, this will reduce the number of allocations.

Reply:
I would be surprised to see that cons allocation is really a bottleneck. Regexp search is most likely the slowest part. But why should we bother with efficiency in the first place? I have never experienced slow indentation or font-lock within R buffers.

Reply:
I would only bother with it if I see it become a problem in practice. We'll see.